File: docs/design/web-ui/v2/selftest.js
<pre id="results"></pre>
<script>
(function(){
  const out=[]; const ok=(name,cond)=>out.push((cond?"PASS ":"FAIL ")+name);
  const key=(k,opts={})=>document.activeElement.dispatchEvent(new KeyboardEvent("keydown",Object.assign({key:k,bubbles:true,cancelable:true},opts)));
  const docKey=(k,opts={})=>document.dispatchEvent(new KeyboardEvent("keydown",Object.assign({key:k,bubbles:true,cancelable:true},opts)));
  const $=s=>document.querySelector(s);
  const input=$("#cmd-input");
  const type=(t)=>{input.focus();input.value=t;input.dispatchEvent(new Event("input",{bubbles:true}));};
  document.body.focus(); docKey("2");
  ok("key 2 opens the checkpoint sheet", !!$(".sheet") && $(".sheet h1").textContent.includes("comp-c"));
  ok("five choices with keys are visible in the sheet foot", document.querySelectorAll(".sheet-foot .choice").length===5);
  docKey("1");
  ok("key 1 opens the confirmation", !!$(".confirm") && $(".confirm h2").textContent==="Approve and run");
  ok("confirmation states the consequence", $(".confirm p").textContent.includes("Pushes kstrl/factory/comp-c"));
  docKey("Escape"); ok("Esc cancels the confirmation and keeps the sheet", !$(".confirm") && !!$(".sheet"));
  docKey("Escape"); ok("Esc closes the sheet", !$(".sheet"));
  docKey("k",{metaKey:true}); ok("cmd+k focuses the command field", document.activeElement===input);
  type("cost?");
  ok("panel opens with the routed answer row", !$("#cmd-panel").classList.contains("hidden") && $("#cmd-panel").textContent.includes("Answer: spend"));
  ok("panel footer cites the recorded latency and tokens", /\d+ ms · [\d,]+ tokens · recorded/.test($(".cmd-foot").textContent));
  key("Enter"); ok("Enter on an ask opens the HUD with the amount", !!$(".hud") && $(".hud .big").textContent.includes("$19.24"));
  type("merge");
  ok("low confidence shows the pick-one list", $("#cmd-panel").textContent.includes("Not sure what you mean") && document.querySelectorAll("#cmd-panel .pbar").length>=3);
  type("approve and merge comp-c"); key("Enter");
  setTimeout(()=>{
    ok("routed decision opens the checkpoint with its confirmation, nothing fired", !!$(".sheet") && !!$(".confirm") && $(".confirm h2").textContent==="Approve and run");
    docKey("Escape"); docKey("Escape");
    type("aprove the merge"); key("Enter"); // codespell:ignore aprove
    ok("decision with no typed target opens the list", !!$(".sheet") && $(".sheet h1").textContent==="Needs you");
    docKey("Escape");
    type("show me failed components this week by cost as a timeline"); key("Enter");
    const fv=$("#fviews .fview"); ok("make a view produces a floating timeline with three marks", !!fv && fv.querySelectorAll(".v-timeline circle").length===3);
    fv.querySelector(".icon-btn").click(); ok("pin moves the view into the dock", !!$("#dock .fview"));
    $("#dock .fview .icon-btn:nth-of-type(2)").click(); ok("close removes the docked view", !$("#dock .fview"));
    type("git push origin main");
    ok("no-match says so", $("#cmd-panel").textContent.includes("Nothing on this page does that"));
    type("word cloud of finding kinds"); key("Enter");
    ok("catalogue gap becomes a component request", !!$("#fviews .request") && $("#fviews .request h4").textContent.includes("word cloud"));
    type("what is the meaning of life");
    ok("unrecorded text is labelled not recorded", $("#cmd-panel").textContent.includes("Not recorded"));
    docKey("Escape"); input.blur(); document.body.focus(); docKey("t"); ok("t switches the theme", document.documentElement.dataset.theme==="light");
    docKey("4"); docKey("r"); ok("retry confirmation names the command", !!$(".confirm") && $(".confirm .cmdline").textContent.includes("ks retry client-commands"));
    $("#results").textContent=out.join("\n")+"\nDONE "+out.filter(x=>x.startsWith("PASS")).length+"/"+out.length;
  },50);
})();
</script>

File: docs/design/web-ui/v2/selftest.py
"""Drive the prototype's keyboard and command flows in headless Chrome and report.

Builds a copy of prototype/index.html with selftest.js appended, loads it
in headless Chrome, and prints one PASS or FAIL line per check from the
page's own DOM. Run: python3 docs/design/web-ui/v2/selftest.py

The checks are the behaviours DESIGN.md promises: number keys open the
Needs you items, a choice key opens a confirmation that states the
consequence, Esc unwinds, the command field routes a recorded phrasing and
labels an unrecorded one, low confidence lists options, a decision with no
typed target opens the list, a made view floats and pins, a catalogue gap
becomes a component request.
"""

# ruff: noqa: E501
from __future__ import annotations

import html as html_mod
import re
import subprocess
import sys
import tempfile
from pathlib import Path

HERE = Path(__file__).resolve().parent
PROTO = HERE / "prototype"
CHROME = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

TEST = (HERE / "selftest.js").read_text(encoding="utf-8")


def main() -> int:
    page = (PROTO / "index.html").read_text(encoding="utf-8")
    for name in ("app.css", "data.js", "replay.js", "app.js"):
        page = page.replace(f'"{name}"', f'"file://{PROTO / name}"')
    page = page.replace("</body>", TEST + "\n</body>")
    with tempfile.TemporaryDirectory() as tmp:
        path = Path(tmp) / "selftest.html"
        path.write_text(page, encoding="utf-8")
        dom = subprocess.run(
            [
                CHROME,
                "--headless=new",
                "--disable-gpu",
                "--virtual-time-budget=6000",
                "--dump-dom",
                f"file://{path}",
            ],
            capture_output=True,
            text=True,
            timeout=120,
            check=False,
        ).stdout
    m = re.search(r'<pre id="results">(.*?)</pre>', dom, re.S)
    if not m:
        print("NO RESULTS: the page did not run the test script")
        return 2
    text = html_mod.unescape(m.group(1))
    print(text)
    return 0 if "FAIL" not in text and "DONE" in text else 1


if __name__ == "__main__":
    sys.exit(main())
